src/ProducerEventProcessor/UploaderCaptureEventCollector.cpp

```diff
@@ -0,0 +1,116 @@
+  // Create an output stream which will be used to convert capture events into parts of a well
+  // formatted capture file. This should never fail.
+  output_stream_ =
+      orbit_capture_file::CaptureFileOutputStream::Create(&capture_data_buffer_stream_);
+}
+
+UploaderCaptureEventCollector::~UploaderCaptureEventCollector() {
+  LOG("Total number of events uploaded: %u", total_uploaded_event_count_);
+  LOG("Total number of bytes uploaded: %u", total_uploaded_data_bytes_);
+  LOG("Total number of write event errors: %u", total_write_error_count_);
+
+  if (total_uploaded_event_count_ > 0) {
+    float average_bytes = static_cast<float>(total_uploaded_data_bytes_) /
+                          static_cast<float>(total_uploaded_event_count_);
+    LOG("Average number of bytes per event: %.2f", average_bytes);
+  }
+}
+
+void UploaderCaptureEventCollector::Stop() {
+  absl::MutexLock lock{&mutex_};
+
+  CHECK(output_stream_ != nullptr);
+  auto close_result = output_stream_->Close();
+  if (close_result.has_error()) {
+    LOG("Error while closing output steam: %s", close_result.error().message());
+  }
+}
+
+void UploaderCaptureEventCollector::AddEvent(ClientCaptureEvent&& event) {
+  {
+    absl::MutexLock lock{&mutex_};
+
+    // The output stream will be closed while processing the capture finish event. Drop events
+    // received after closing the output stream.
+    CHECK(output_stream_ != nullptr);
+    if (!output_stream_->IsOpen()) return;
+
+    auto write_result = output_stream_->WriteCaptureEvent(event);
+    if (write_result.has_error()) {
+      total_write_error_count_++;
+      return;
+    }
+    buffered_event_count_++;
+    buffered_event_bytes_ += event.ByteSizeLong();
+  }
+
+  // Close output stream after processing the capture finish event.
+  if (event.event_case() == ClientCaptureEvent::kCaptureFinished) Stop();
+}
+
+DataReadiness UploaderCaptureEventCollector::GetDataReadiness() const {
+  absl::MutexLock lock(&mutex_);
+
+  if (!capture_data_to_upload_.empty()) {
+    return DataReadiness::kHasData;
+
+  } else if (output_stream_->IsOpen() || buffered_event_bytes_ > 0) {
+    return DataReadiness::kWaitingForData;
+
+  } else {
+    return DataReadiness::kEndOfData;
+  }
+}
+
+void UploaderCaptureEventCollector::RefreshUploadDataBuffer() {
+  // Clear capture_data_to_upload_ immediately.
+  capture_data_to_upload_.clear();
+
+  // Refill capture_data_to_upload_ when there is enough data to upload.
+  mutex_.LockWhen(absl::Condition(
+      +[](UploaderCaptureEventCollector* self) ABSL_EXCLUSIVE_LOCKS_REQUIRED(self->mutex_) {
+        constexpr int kUploadEventCountInterval = 5000;
+        return self->buffered_event_count_ >= kUploadEventCountInterval ||
+               !self->output_stream_->IsOpen();
+      },
+      this));
+
+  ORBIT_INT("Number of CaptureEvents to upload", buffered_event_count_);
+  ORBIT_INT("Bytes of CaptureEvents to upload", buffered_event_bytes_);
```

Comment:
These are `size_t`, so `ORBIT_UINT64` is more appropriate.

Reply:
Fixed now. Thanks!